File: lib/data/models/match_data.dart
import 'dart:collection';
import 'dart:math';

enum MatchResult { win, lose, draw }

class LocationPoint {
  final double x;
  final double y;
  final DateTime timestamp;
  final double speedKmh;

  const LocationPoint({
    required this.x,
    required this.y,
    required this.timestamp,
    this.speedKmh = 0,
  });
}

class MatchStats {
  final double totalDistanceKm;
  final int averageHeartRate;
  final int maxHeartRate;
  final double maxSpeedKmh;
  final double averageSpeedKmh;
  final int sprintCount;
  final int caloriesBurned;
  final double sprintDistanceKm;

  const MatchStats({
    this.totalDistanceKm = 0,
    this.averageHeartRate = 0,
    this.maxHeartRate = 0,
    this.maxSpeedKmh = 0,
    this.averageSpeedKmh = 0,
    this.sprintCount = 0,
    this.caloriesBurned = 0,
    this.sprintDistanceKm = 0,
  });
}

class FieldSize {
  final double lengthMeters;
  final double widthMeters;

  const FieldSize({
    this.lengthMeters = 105,
    this.widthMeters = 68,
  });

  double get area => lengthMeters * widthMeters;
}

class MatchData {
  final String id;
  final DateTime date;
  final int durationMinutes;
  final int myScore;
  final int opponentScore;
  final MatchResult result;
  final MatchStats stats;
  final List<LocationPoint> _locationHistory;
  final FieldSize? fieldSize;
  final String? matchName;

  /// 외부 수정 불가능한 locationHistory 반환
  List<LocationPoint> get locationHistory =>
      UnmodifiableListView(_locationHistory);

  MatchData({
    required this.id,
    required this.date,
    required this.durationMinutes,
    required this.myScore,
    required this.opponentScore,
    required this.result,
    required this.stats,
    List<LocationPoint> locationHistory = const [],
    this.fieldSize,
    this.matchName,
  }) : _locationHistory = List.unmodifiable(locationHistory);

  /// 지정된 시간 범위로 경기 데이터를 자르고 통계를 재계산합니다.
  MatchData trim({required DateTime start, required DateTime end}) {
    // 1. 유효성 검사: 날짜 순서 확인
    if (start.isAfter(end)) {
      throw ArgumentError('Start time must be before end time');
    }

    // 2. 데이터 필터링
    final filteredLocations = locationHistory.where((p) {
      return (p.timestamp.isAfter(start) || p.timestamp.isAtSameMomentAs(start)) &&
             (p.timestamp.isBefore(end) || p.timestamp.isAtSameMomentAs(end));
    }).toList();

    // 3. 데이터 무결성 검사 (점이 너무 적으면 원본 반환 또는 에러)
    if (filteredLocations.length < 2) {
      // 데이터가 너무 적어 통계 계산이 불가능할 경우 원본을 반환하거나 예외 처리
      // 여기서는 안전하게 기존 데이터를 유지합니다.
      return this;
    }

    // 4. 통계 재계산
    double newTotalDistMeters = 0;
    double maxSpeed = 0;
    
    // 거리 및 속도 계산
    for (int i = 0; i < filteredLocations.length; i++) {
      final p = filteredLocations[i];

      // Max Speed
      if (p.speedKmh > maxSpeed) {
        maxSpeed = p.speedKmh;
      }

      // Total Distance
      if (i > 0) {
        final prev = filteredLocations[i - 1];
        // 유클리드 거리 계산 (미터 단위 가정)
        final dist = sqrt(pow(p.x - prev.x, 2) + pow(p.y - prev.y, 2));
        newTotalDistMeters += dist;
      }
    }

    // 새 지속 시간 (분)
    final newDurationMinutes = filteredLocations.last.timestamp
        .difference(filteredLocations.first.timestamp)
        .inMinutes;

    // 평균 속도 (km/h) = 거리(km) / 시간(h)
    double newAvgSpeed = 0;
    if (newDurationMinutes > 0) {
      newAvgSpeed = (newTotalDistMeters / 1000) / (newDurationMinutes / 60);
    } else if (filteredLocations.isNotEmpty) {
      // 시간이 0분으로 잡힐 정도로 짧은 경우, 점들의 평균 속도로 대체
      double sumSpeed = filteredLocations.fold(0, (sum, p) => sum + p.speedKmh);
      newAvgSpeed = sumSpeed / filteredLocations.length;
    }

    // 새 통계 객체 생성
    // 참고: sprintCount, caloriesBurned 등은 단순 계산이 어려우므로 
    // 시간 비율로 줄이거나 기존 값을 유지해야 함. 여기서는 보수적으로 비율 감소 적용.
    double ratio = 1.0;
    if (durationMinutes > 0) {
      ratio = newDurationMinutes / durationMinutes;
    }
    
    final newStats = MatchStats(
      totalDistanceKm: newTotalDistMeters / 1000,
      averageHeartRate: stats.averageHeartRate, // 유지
      maxHeartRate: stats.maxHeartRate, // 유지
      maxSpeedKmh: maxSpeed,
      averageSpeedKmh: newAvgSpeed,
      sprintCount: (stats.sprintCount * ratio).round(), // 비율로 감소
      caloriesBurned: (stats.caloriesBurned * ratio).round(), // 비율로 감소
      sprintDistanceKm: stats.sprintDistanceKm * ratio, // 비율로 감소
    );

    // 5. 새 MatchData 객체 반환
    return MatchData(
      id: id,
      date: filteredLocations.first.timestamp, // 시작 시간 변경
      durationMinutes: newDurationMinutes,
      myScore: myScore,
      opponentScore: opponentScore,
      result: result,
      stats: newStats,
      locationHistory: filteredLocations,
      fieldSize: fieldSize,
      matchName: matchName,
    );
  }

  String get resultText {
    switch (result) {
      case MatchResult.win:
        return '승리';
      case MatchResult.lose:
        return '패배';
      case MatchResult.draw:
        return '무승부';
    }
  }

  String get scoreText => '$myScore : $opponentScore';
}

class WeeklyStats {
  final int totalCalories;
  final double totalDistanceKm;
  final int matchCount;
  final int totalSprints;
  final double maxSpeedKmh;

  const WeeklyStats({
    this.totalCalories = 0,
    this.totalDistanceKm = 0,
    this.matchCount = 0,
    this.totalSprints = 0,
    this.maxSpeedKmh = 0,
  });
}

File: Kickoff/Models/MatchData.swift
import Foundation

enum MatchResult: String, Codable {
    case win, lose, draw

    var text: String {
        switch self {
        case .win: return "승리"
        case .lose: return "패배"
        case .draw: return "무승부"
        }
    }

    var emoji: String {
        switch self {
        case .win: return "W"
        case .lose: return "L"
        case .draw: return "D"
        }
    }
}

struct FieldSize: Codable {
    var lengthMeters: Double
    var widthMeters: Double

    init(lengthMeters: Double = 105, widthMeters: Double = 68) {
        self.lengthMeters = lengthMeters
        self.widthMeters = widthMeters
    }

    var area: Double { lengthMeters * widthMeters }
}

struct MatchData: Codable, Identifiable {
    let id: String
    let date: Date
    var durationMinutes: Int
    var myScore: Int
    var opponentScore: Int
    var result: MatchResult
    var stats: MatchStats
    var locationHistory: [LocationPoint]
    var fieldSize: FieldSize?
    var matchName: String?

    init(
        id: String = UUID().uuidString,
        date: Date = Date(),
        durationMinutes: Int = 0,
        myScore: Int = 0,
        opponentScore: Int = 0,
        result: MatchResult = .draw,
        stats: MatchStats = .empty,
        locationHistory: [LocationPoint] = [],
        fieldSize: FieldSize? = nil,
        matchName: String? = nil
    ) {
        self.id = id
        self.date = date
        self.durationMinutes = durationMinutes
        self.myScore = myScore
        self.opponentScore = opponentScore
        self.result = result
        self.stats = stats
        self.locationHistory = locationHistory
        self.fieldSize = fieldSize
        self.matchName = matchName
    }

    var resultText: String { result.text }
    var scoreText: String { "\(myScore) : \(opponentScore)" }

    var formattedDate: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "ko_KR")
        formatter.dateFormat = "M월 d일 (E)"
        return formatter.string(from: date)
    }

    var formattedDuration: String {
        let hours = durationMinutes / 60
        let minutes = durationMinutes % 60
        if hours > 0 {
            return "\(hours)시간 \(minutes)분"
        }
        return "\(minutes)분"
    }

    static let sample = MatchData(
        id: "sample-1",
        date: Date(),
        durationMinutes: 90,
        myScore: 3,
        opponentScore: 1,
        result: .win,
        stats: MatchStats(
            totalDistanceKm: 8.5,
            averageHeartRate: 145,
            maxHeartRate: 185,
            maxSpeedKmh: 28.3,
            averageSpeedKmh: 6.2,
            sprintCount: 24,
            caloriesBurned: 720,
            sprintDistanceKm: 1.8
        ),
        locationHistory: LocationPoint.samplePoints,
        fieldSize: FieldSize(),
        matchName: "주말 리그"
    )
}

extension LocationPoint {
    static var samplePoints: [LocationPoint] {
        var points: [LocationPoint] = []
        let now = Date()
        for i in 0..<200 {
            let t = Double(i) / 200.0
            let x = 0.3 + 0.4 * sin(t * .pi * 4) * 0.5 + Double.random(in: -0.1...0.1)
            let y = 0.2 + 0.6 * t + Double.random(in: -0.1...0.1)
            let speed = Double.random(in: 2...25)
            points.append(LocationPoint(
                x: min(max(x, 0), 1),
                y: min(max(y, 0), 1),
                timestamp: now.addingTimeInterval(TimeInterval(i) * 27),
                speedKmh: speed
            ))
        }
        return points
    }
}
